Mathlib/Data/Polynomial/Smeval.lean

```diff
@@ -133,17 +138,158 @@ theorem leval_coe_eq_smeval {R : Type*} [Semiring R] (r : R) :
+
+theorem smeval_mul_X_pow : ∀(n : ℕ), (p * X^n).smeval x = p.smeval x * x^n
+  | 0 => by
+    simp only [npow_zero, mul_one]
+  | n + 1 => by
+    rw [npow_add, ← mul_assoc, npow_one, smeval_mul_X, smeval_mul_X_pow n, npow_add,
+      ← smeval_assoc_X_pow, npow_one]
+
+theorem smeval_X_pow_assoc (m n : ℕ) :
+    x ^ m * x ^ n * p.smeval x = x ^ m * (x ^ n * p.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [smeval_add, ph, qh, mul_add]
+  | h_monomial n a =>
+    simp only [smeval_monomial, mul_smul_comm, npow_mul_assoc]
+
+theorem smeval_X_pow_mul : ∀(n : ℕ), (X^n * p).smeval x = x^n * p.smeval x
+  | 0 => by
+    simp [npow_zero, one_mul]
+  | n + 1 => by
+    rw [add_comm, npow_add, mul_assoc, npow_one, smeval_X_mul, smeval_X_pow_mul n, npow_add,
+      smeval_X_pow_assoc, npow_one]
+
+theorem smeval_C_mul : (C r * p).smeval x = r • p.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [mul_add, smeval_add, ph, qh, smul_add]
+  | h_monomial n b =>
+    simp only [C_mul_monomial, smeval_monomial, mul_smul]
+
+theorem smeval_monomial_mul (n : ℕ) :
+    (monomial n r * p).smeval x = r • (x ^ n * p.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+    rw [← @C_mul_X_pow_eq_monomial, mul_assoc, smeval_C_mul, smeval_X_pow_mul, smeval_add]
+  | h_monomial n a =>
+    rw [smeval_monomial, monomial_mul_monomial, smeval_monomial, npow_add, mul_smul, mul_smul_comm]
+
+theorem smeval_mul : (p * q).smeval x  = p.smeval x * q.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+  | h_monomial n a =>
+    simp only [smeval_monomial, smeval_C_mul, smeval_mul_X_pow, smeval_monomial_mul, smul_mul_assoc]
+
+theorem smeval_pow : ∀(n : ℕ), (p^n).smeval x = (p.smeval x)^n
+  | 0 => by
+    simp only [npow_zero, smeval_one, one_smul]
+  | n + 1 => by
+    rw [npow_add, smeval_mul, smeval_pow n, pow_one, npow_add, npow_one]
+
+theorem smeval_comp : (p.comp q).smeval x  = p.smeval (q.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp [add_comp, hr, hs, smeval_add]
+  | h_monomial n a =>
+    simp [smeval_monomial, smeval_C_mul, smeval_pow]
+
+end NatPowAssoc
+
+section Algebra
 
 theorem aeval_coe_eq_smeval {R : Type*} [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
-    (x : S) : ⇑(aeval x) = fun p => @smeval S _ _ x R _ _ p := by
+    (x : S) : ⇑(aeval x) = fun p => @smeval R _ p S _ _ _ x := by
   refine funext ?_
   intro
   rw [aeval_def, eval₂_def, Algebra.algebraMap_eq_smul_one', smeval_def]
   simp only [Algebra.smul_mul_assoc, one_mul]
   exact rfl
 
-end Comparisons
+/-- `Polynomial.smeval` as an algebra map. -/
+def smeval.algebraMap {R : Type*} [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
```

Comment:
```suggestion
def smeval.algebraMap (R : Type*) [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
```
I make this suggestion because it enables you to remove an `@` later on. Experience will end up telling you whether you want `R` explicit or implicit.